### PR TITLE
BufferedInputBase pre-reserves data to DEFAULT_BUFFER_SIZE

### PR DESCRIPTION
**Pull request by terencezl**

### CODE REVIEW

**Comment by terencezl** [comment]
https://github.com/RaRe-Technologies/smart_open/issues/348

File: smart_open/s3.py [hunk 1]
```diff
@@ -281,7 +281,8 @@ def read(self, size=-1):
         # Fill our buffer to the required size.
         #
         # logger.debug('filling %r byte-long buffer up to %r bytes', len(self._buffer), size)
-        self._fill_buffer(size)
+        fill_size = size if size > DEFAULT_BUFFER_SIZE else DEFAULT_BUFFER_SIZE
```
Comment (mpenkov) [inline smart_open/s3.py]:
```suggestion
        fill_size = max(size, DEFAULT_BUFFER_SIZE)
```

**Comment by mpenkov** [comment]
@terencezl Are you able to finish this PR? It's almost done.

If not, can you please enable "allow edits by maintainers"? If you're unable to finish, I will push it over the line.

**Comment by terencezl** [comment]
On the right side bar it seems already enabled. I did not see the difference in performance when connected to an on-premise object store, and am not sure how to perform this integration test in a general wat.

On second thought if this is useful, we can even use the user provided buffer size instead of just the default value.

File: smart_open/s3.py (hunk 1 above)
Comment (mpenkov) [reply smart_open/s3.py]: Could you please apply this change?

I get "You do not have permission to push to this repository."

**Comment by mpenkov** [comment]
Closing as stale

**Pull request closed without merging**